File: Tutorials/Codes/Modern_Control_Course/Chapter2/Lesson4/Chapter2_Lesson4.cpp
#include <iostream>
#include <Eigen/Dense>

Eigen::MatrixXd similarityTransform(const Eigen::MatrixXd& A,
                                    const Eigen::MatrixXd& T) {
    // Prefer solving linear systems rather than forming inverse explicitly:
    // A_tilde = T^{-1} A T  =>  solve(T, (A*T))
    Eigen::MatrixXd AT = A * T;
    Eigen::MatrixXd A_tilde = T.fullPivLu().solve(AT);
    return A_tilde;
}

int main() {
    Eigen::MatrixXd A(2,2);
    A << 2.0, 1.0,
         0.0, 3.0;

    Eigen::MatrixXd T(2,2);
    T << 1.0, 1.0,
         0.0, 1.0;

    Eigen::MatrixXd A_tilde = similarityTransform(A, T);

    std::cout << "A:\n" << A << "\n\n";
    std::cout << "A_tilde:\n" << A_tilde << "\n\n";
    std::cout << "trace(A) = " << A.trace()
              << ", trace(A_tilde) = " << A_tilde.trace() << "\n";
    std::cout << "det(A) = " << A.determinant()
              << ", det(A_tilde) = " << A_tilde.determinant() << "\n";
    return 0;
}
      

File: Tutorials/Codes/Modern_Control_Course/Chapter2/Lesson4/Chapter2_Lesson4.py
import numpy as np

def gauss_jordan_inverse(A):
    A = np.array(A, dtype=float)
    n = A.shape[0]
    I = np.eye(n)
    Aug = np.hstack([A, I])

    # Forward elimination with partial pivoting
    for col in range(n):
        pivot = np.argmax(np.abs(Aug[col:, col])) + col
        Aug[[col, pivot]] = Aug[[pivot, col]]

        piv = Aug[col, col]
        if np.isclose(piv, 0.0):
            raise ValueError("Matrix is singular or nearly singular.")

        Aug[col, :] = Aug[col, :] / piv

        for row in range(n):
            if row == col:
                continue
            factor = Aug[row, col]
            Aug[row, :] = Aug[row, :] - factor * Aug[col, :]

    return Aug[:, n:]

def similarity_transform(A, T):
    # Prefer solving rather than explicit inverse in larger problems;
    # kept explicit here for pedagogy.
    Tinv = gauss_jordan_inverse(T)
    return Tinv @ A @ T

# Example
A = np.array([[2.0, 1.0],
              [0.0, 3.0]])
T = np.array([[1.0, 1.0],
              [0.0, 1.0]])

A_tilde = similarity_transform(A, T)

# Invariants check (eigenvalues should match)
eigA = np.linalg.eigvals(A)
eigAt = np.linalg.eigvals(A_tilde)

print("A =", A)
print("A_tilde =", A_tilde)
print("eig(A) =", eigA)
print("eig(A_tilde) =", eigAt)
print("trace equal:", np.isclose(np.trace(A), np.trace(A_tilde)))
print("det equal:", np.isclose(np.linalg.det(A), np.linalg.det(A_tilde)))
      
